File: leetcode/Rotate Array.py
class Solution:

    def bruteForce(self, nums: List[int], k: int) -> None:
        n = len(nums)
        k = k % n

        for i in range(k):
            prev = nums[-1]
            for j in range(n):
                nums[j], prev = prev, nums[j]

    def cyclicReplacement(self, nums: List[int], k: int) -> None:
        n = len(nums)
        k %= n

        start, count = 0, 0

        # SELF_FEEDBACK: come up with a terminal point!
        while count < n:
            # do while
            currIdx, currVal = start, nums[start]
            while True:
                nextIdx = (currIdx + k) % n

                # replace
                tempVal = nums[nextIdx]
                nums[nextIdx] = currVal

                currVal = tempVal
                currIdx = nextIdx
                count += 1

                if currIdx == start:
                    break

            start += 1

    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.

        time: O(n)
        space: O(1)
        """
        self.bruteForce(nums, k)
        # self.cyclicReplacement(nums, k)

File: java-kotlin/src/leetcode/java/RotateArray.java
package leetcode.java;

public class RotateArray {

  public void rotate(int[] nums, int k) {
    rotateWithExtraSpace(nums, k);
    // simulateRotating(nums, k);
  }

  // time: O(N), space: O(N)
  private void rotateWithExtraSpace(int[] nums, int k) {
    k = k % nums.length;
    int[] temp = new int[nums.length];

    for (int i = 0; i < nums.length; i++) {
      temp[(i + k) % nums.length] = nums[i];
    }

    for (int i = 0; i < nums.length; i++) {
      nums[i] = temp[i];
    }
  }

  // time: O(N*k), space: O(1)
  private void simulateRotating(int[] nums, int k) {
    k = k % nums.length;

    for (int i = 0; i < k; i++) {
      rotateOneStep(nums);
    }
  }

  private void rotateOneStep(int[] nums) {
    int prev = nums[nums.length - 1];
    for (int i = 0; i < nums.length; i++) {
      int temp = nums[i];
      nums[i] = prev;
      prev = temp;
    }
  }
}
